odb_extractor_atoms.py

```diff
@@ -59,98 +223,297 @@ def select_obsclass(classes):
-                    sequence[ii + 1]['telescope:q']) == float(sequence[ii + 2]['telescope:q']):
-                nabba = 3
+        # A change in exposure time or coadds is a new atom for science exposures
+        if step['observe:observeType'].upper() not in ['FLAT', 'ARC', 'DARK', 'BIAS']:
+            if observe_class.upper() == 'SCIENCE' and \
+                    ii > 0 and (exptimes[ii] != exptimes[ii_prevobj] or coadds[ii] != coadds[ii_prevobj]):
                 nextatom = True
-                print('Atom for ABBA')
-        else:
-            nabba -= 1
-
+                atomstr += 'exposure time change, '
+
+            # Offsets - a new offset pattern is a new atom
+            #         print('npattern: ', npattern)
+            if not (offset_lag == 0 and exptime_groups == True):
+                # For NIR imaging, need to have at least two offset positions if no repeating pattern
+                # New atom after every 2nd offset (noffsets is odd)
+                if mode == 'imaging' and offset_lag == 0 and all([w > 1.0 for w in config['wavelength']]):
+                    if ii == 0:
+                        noffsets += 1
+                    else:
+                        if poffsets[ii] != poffsets[ii_prevobj] or qoffsets[ii] != qoffsets[ii_prevobj]:
+                            noffsets += 1
+                    if noffsets % 2 == 1:
+                        nextatom = True
+                        atomstr += 'offset pattern'
+                else:
+                    npattern -= 1
+                    if npattern < 0:
+                        nextatom = True
+                        atomstr += 'offset pattern'
+                        npattern = offset_lag - 1
+
+            ii_prevobj = ii
+
+        # New atom?
         if nextatom:
-            # Get class and qastate for previous atom
+            # Get class, qastate, guiding for previous atom
             if natom > 0:
-                print(qastates)
-                atoms[-1]['qastate'] = select_qastate(qastates)
-                print(classes)
+                atoms[-1]['qa_state'] = select_qastate(qastates)
+                if atoms[-1]['qa_state'] != 'NONE':
+                    atoms[-1]['observed'] = True
                 atoms[-1]['class'] = select_obsclass(classes)
-
-            # New atom
+                if verbose:
+                    print('QA states: ', qastates, file=fid)
+                    print('Classes: ', classes, file=fid)
+                    print('Guiding: ', guiding, file=fid)
+                atoms[-1]['guide_state'] = any(guiding)
+                atoms[-1]['wavelength'] = config['wavelength'][ii]
+                atoms[-1]['required_resources']['inst'] = config['inst']
+                atoms[-1]['required_resources']['filter'] = config['filter'][ii]
+                atoms[-1]['required_resources']['disperser'] = config['disperser'][ii]
+                atoms[-1]['required_resources']['fpu'] = config['fpu'][ii]
```

Comment:
This I find particularly confusing: why are we modifying the last atom in the list instead of creating a new one and appending it? I think that may be equivalent to what is happening with the `atoms.append` below, but it seems this could be done in a way that is clearer?